extension.ts

```diff
@@ -0,0 +1,374 @@
+): SearchProvider => ({
+  id: Self.uuid,
+  canLaunchSearch: true,
+  isRemoteProvider: false,
+  appInfo: vscode,
+  getInitialResultSet: (terms, callback): void =>
+    callback(
+      findMatchingItems(
+        Array.from(items.values()),
+        terms,
+        getEnabledKinds(settings)
+      )
+    ),
+  getSubsearchResultSet: (current, terms, callback): void =>
+    callback(
+      findMatchingItems(
+        lookupRecentItems(items, current),
+        terms,
+        getEnabledKinds(settings)
+      )
+    ),
+  getResultMetas: (ids, callback): void =>
+    callback(lookupRecentItems(items, ids).map(resultMetaOfRecentItem(vscode))),
+  launchSearch: (): void => launchVSCodeInShell(vscode),
+  activateResult: (id): void => {
+    const item = items.get(id);
+    if (item) {
+      launchVSCodeInShell(vscode, [item.file]);
+    }
+  },
+  filterResults: (results, max): string[] => results.slice(0, max)
+});
+
+/**
+ * Create a recent item from a URI.
+ *
+ * @param kind The kind of the new item
+ * @returns The recent item for the URI
+ */
+const createRecentItem = (kind: RecentItemKind) => (
+  uri: string
+): RecentItem => {
+  const file = Gio.File.new_for_uri(uri);
+  return {
+    id: `vscode-search-provider-${uri}`,
+    // TODO: Check the outcome of get_basename on file URIs
+    name: file.get_basename() || `<unnamed ${kind}>`,
+    // TODO: Test what this parse name thing actually is
+    readablePath: file.get_parse_name(),
+    kind,
+    file
+  };
+};
+
+/**
+ * Find recent items from VSCode.
+ *
+ * @param configDirectoryName The name of the config directory of the code app
+ * @returns A promise with recent items
+ */
+const findVSCodeRecentItems = (
+  configDirectoryName: string
+): Promise<RecentItems> =>
+  new Promise(resolve => {
+    // TODO: Use async load contents for async IO, to avoid blocking the shell
+    const contents = Gio.File.new_for_path(GLib.get_user_config_dir())
+      .get_child(configDirectoryName)
+      .get_child("storage.json")
+      .load_contents(null)[1];
+
+    const storage = JSON.parse(ByteArray.toString(contents));
+
+    const recentWorkspaceURIs: string[] =
+      storage.openedPathsList.workspaces3 ||
+      storage.openedPathsList.workspaces2 ||
+      storage.openedPathsList.workspaces ||
+      [];
+    const recentFileURIs: string[] =
+      storage.openedPathsList.files2 || storage.openedPathsList.files || [];
+
+    const recentItems = recentWorkspaceURIs
+      .map(createRecentItem("workspace"))
+      .concat(recentFileURIs.map(createRecentItem("file")));
+
+    resolve(new Map(recentItems.map(item => [item.id, item])));
+  });
+
+/**
+ * The registered provider
+ */
+// eslint-disable-next-line immutable/no-let
+let registeredProvider: "unregistered" | "registering" | SearchProvider =
+  "unregistered";
+
+/**
+ * Initialize this extension.
+ */
+// eslint-disable-next-line @typescript-eslint/no-unused-vars,@typescript-eslint/no-empty-function
+function init(): void {}
```

Comment:
Can we export init, enable and disable, rather than disabling the linter? Also makes it a bit clearer that they are used externally.
(I know I didn't previously as well 😄  )

Reply:
Export as in `export function`?  I tried this, but it makes tsc output `extension.js` a CommonJS module; specifically `extension.js` then wants the `exports` global in scope.  I didn't try to use this with GJS or Gnome shell, but I just don't think that Gnome Shell provides a commonjs runtime to extensions – if it did they'd surely document that somewhere, wouldn't they?

I also tried to disable the `no-unused-vars` rule entirely and rely on tsc, but tsc only detects unused locals and arguments, but not unused globals.

As far as I can see, selectively disabling the rule for the three extension entry points is the best we can do.

Reply:
Guess one would need to mess with tsconfig to get it to work properly, yeah..
So no worries on that